🌱 Add Node related condition to Machine conditions

sedefsavas opened the pull request:
**What this PR does / why we need it**:
This PR adds a condition for Node health to Machine conditions that will be managed by Machine controller.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api/issues/3138

sedefsavas commented:
cc @fabriziopandini

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+					"failed to find the Node owned by Machine %q in namespace %q", machine.Name, machine.Namespace)
+			}
+
+			logger.Error(err, "Failed to assign NodeRef")
+			r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
```
We should probably keep logging and recording the event regardless of the error, like it was before. This behavior change might be worthwhile to investigate in v1alpha4

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+// Return true if all Node Conditions are false other than "Ready" condition.
+// "Ready" being false does not mean there is a problem with the Node, may be missing CNI.
+func checkNodeConditions(node *apicorev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type != apicorev1.NodeReady {
+			if condition.Status == apicorev1.ConditionTrue {
+				return false
+			}
+		}
+	}
```
Should we check specific conditions, instead of ignoring `Ready` and checking all the others?

vincepri replied:
Also we might want to add a bit more checks for control plane nodes?

sedefsavas replied:
> Also we might want to add a bit more checks for control plane nodes?



I will have another PR that adds control-plane node conditions to the machines in `KubeadmControlPlane` controller. Since the owner of control planes machines are KCP, I think that will be a better place to add control-plane specific conditions there.

fabriziopandini replied:
+1 to check a list of conditions. I will include ready thought.

sedefsavas replied:
I added the ready condition check too, but do we consider a Node unhealthy if it is not ready due to missing CNI?

vincepri replied:
Probably not, although we don't have a way to detect if CNI has been installed or not. Checking `Ready` might also be overstepping our responsibilities with MachineHealthCheck, which already has timeouts for node ready

fabriziopandini replied:
My concern is the user perspective, and what I want to avoid is having the users being confused by the machine reporting the node is healthy, while the node itself reports being not ready
However, I plan to test this as soon as draft is lifted from the PR.

fabriziopandini replied:
Let's tackle this in a separated PR

fabriziopandini commented on api/v1alpha3/condition_consts.go:
```diff
@@ -109,9 +109,6 @@ const (
 	MachineHasFailureReason = "MachineHasFailure"
 
-	// NodeNotFoundReason is the reason used when a machine's node has previously been observed but is now gone.
-	NodeNotFoundReason = "NodeNotFound"
```
Should we deprecate instead of deleting?

sedefsavas replied:
It is moved below, not deleted.

fabriziopandini commented on api/v1alpha3/condition_consts.go:
```diff
@@ -120,10 +117,48 @@ const (
 	WaitingForRemediationReason = "WaitingForRemediation"
 )
+
+// Conditions and condition Reasons for the Machine's Node object
```
nit. I think we can avoid this separator because the following conditions are still machine conditions

fabriziopandini commented on api/v1alpha3/condition_consts.go:
```diff
@@ -120,10 +117,48 @@ const (
+	NodeConditionsFailedReason = "NodeConditionsFailed"
+)
+
+// Common Pod-related Condition Reasons used by Pod-related Conditions.
```
```suggestion
// Common Reasons that applyes to MachineKubeAPIServerHealthyCondition, MachineKubeControllerManagerHealthyCondition, MachineKubeControllerManagerHealthyCondition.
```
If I'm not wrong, those reason should apply to the etcd pod as well

fabriziopandini replied:
Also if I'm not wrong we are not declaring ^^ conditions anywhere

sedefsavas replied:
You are right, we are using them here: https://github.com/kubernetes-sigs/cluster-api/pull/3674
Removed.

fabriziopandini commented on controllers/machine_controller.go:
```diff
@@ -252,7 +252,7 @@ func (r *MachineReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
 		r.reconcileBootstrap(ctx, cluster, m),
 		r.reconcileInfrastructure(ctx, cluster, m),
-		r.reconcileNodeRef(ctx, cluster, m),
+		r.reconcileNode(ctx, cluster, m),
```
should we rename `controllers/machine_controller_noderef.go` as well?

fabriziopandini reviewed (commented):
first pass

sedefsavas commented on controllers/machine_controller_noderef.go:
```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
```
When the machine NodeRef is just set in this reconcile, do we want to go ahead and set node related conditions below or just `return nil` like I did here and in the next reconcile, set them?

vincepri replied:
Seems like there are no extra calls being made after this line of code, so I'd say you can remove the return and instead add a comment explaining why we're proceeding with the other operations?

sedefsavas commented:
Unit test fails because CacheTracker panics in the tests that doesn't use testEnv. Probably will need to use testEnv for the tests using CacheTracker.

sedefsavas commented on controllers/machine_controller_test.go:
```diff
@@ -40,6 +41,107 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+var _ = Describe("Machine Reconciler", func() {
```
We are trying to avoid using ginkgo tests going forward but for this particular case:
To use testenv with go tests in ./controllers, we need to make changes in `suite_test.go` similar to `controlplane/kubeadm/internal/suite_test.go`. 
Currently all tests that use testenv in ./controllers package are ginkgo tests, so I think we will need to change them all together with the `suite_test.go`.

vincepri commented on controllers/machine_controller_phases_test.go:
```diff
@@ -104,7 +107,7 @@ var _ = Describe("Reconcile Machine Phases", func() {
 
 	BeforeEach(func() {
-		defaultKubeconfigSecret = kubeconfig.GenerateSecret(defaultCluster, kubeconfig.FromEnvTestConfig(testEnv.Config, defaultCluster))
+		defaultKubeconfigSecret = kubeconfig.GenerateSecret(defaultCluster, kubeconfig.FromEnvTestConfig(&rest.Config{}, defaultCluster))
```
What does this do?

sedefsavas replied:
This test doesn't use envtest so it actually doesn't need envtest config here.

vincepri commented on api/v1alpha3/condition_consts.go:
```diff
@@ -120,10 +117,28 @@ const (
 
 const (
-	// MachineOwnerRemediatedCondition is set on machines that have failed a healthcheck by the MachineHealthCheck controller.
+	// MachineOwnerRemediatedCondition is set on machines that have failed a healthcheck by the Machine's owner controller.
```
Revert this? The older one was correct, MHC sets the condition

sedefsavas replied:
I was trying to cover the case where we don't use MHC for KCP remediation, but currently this is not the case. Reverted.

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -76,13 +77,50 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
-func (r *MachineReconciler) getNodeReference(c client.Reader, providerID *noderefutil.ProviderID) (*apicorev1.ObjectReference, error) {
+// checkNodeConditions checks if a Node is healthy and does not have any semantically negative Conditions.
+// Returns true if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func checkNodeConditions(node *apicorev1.Node) bool {
```
```suggestion
func hasHealthyNodeConditions(node *apicorev1.Node) bool {
```

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -76,13 +77,50 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+// Returns true if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func checkNodeConditions(node *apicorev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == apicorev1.NodeMemoryPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodeDiskPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodePIDPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodeReady && condition.Status == apicorev1.ConditionFalse {
+			return false
+		}
```
Can we have a switch case here instead of listing if conditions one by one?

As an example:
```suggestion
		switch condition.Type {
			case apicorev1.NodeMemoryPressure, NodeDiskPressure, NodePIDPressure:
				return condition.Status != apicorev1.ConditionTrue
			[...] and so on [...]
		}
```

vincepri commented on controllers/machine_controller_node_test.go:
```diff
@@ -108,7 +108,7 @@ func TestGetNodeReference(t *testing.T) {
 			gt.Expect(err).NotTo(HaveOccurred(), "Expected no error parsing provider id %q, got %v", test.providerID, err)
 
-			reference, err := r.getNodeReference(client, providerID)
+			reference, err := r.getNode(client, providerID)
```
```suggestion
			node, err := r.getNode(client, providerID)
```

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -76,13 +77,50 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if !checkNodeConditions(node) {
```
Should we return an error instead of a boolean? Maybe explaining which node condition (or actually, a set of conditions, we can use an aggregate error for that) is causing problems

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -76,13 +77,50 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+
```
```suggestion

	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
```

vincepri commented:
/milestone v0.4.0

sedefsavas commented:
/test pull-cluster-api-test

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
```
what about setting MachineNodeHealthyCondition=NodeProvisioning in the else branch?

sedefsavas replied:
It is set below. Here node is not found, but we want to set `MachineNodeHealthyCondition=NodeProvisioning` when noderef is nil whether or not node is found.

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+
+// hasHealthyNodeConditions checks if a Node is healthy and does not have any semantically negative Conditions.
+// Returns error if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func hasHealthyNodeConditions(node *apicorev1.Node) error {
```
I personally found the signature of this func a little bit confusing, mostly because it is using an errors & kerrors.NewAggregate to compose a message.
What about 
```suggestion
func summarizeNodeConditions(node *apicorev1.Node) (status corev1.ConditionStatus, message string) {
```

sedefsavas replied:
With this signature, we can only relay the first condition that fails. If we want to show all the failed conditions, error list makes sense in the sense that these are all errors on the Node side. Or I can use a string slice?

fabriziopandini replied:
The return message could be the concatenation of the failing conditions. 
I would avoid exposing a slice of messages

sedefsavas replied:
I changed the signature like this but status has different meanings for different conditions, so to be consistent, returning false if a condition is false semantically.

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+			break
+		}
+	}
+	return nil, ErrNodeNotFound
```
wondering if `return nil, nil` is equally expressive to represent node not found (thus avoiding the need of `ErrorType`)

fabriziopandini reviewed (commented):
Another pass
I'm going to test this together with all the other conditions PR in flight

vincepri commented on api/v1alpha3/condition_consts.go:
```diff
@@ -127,3 +124,21 @@ const (
+const (
+	// MachineNodeHealthyCondition provides info about the operational state of the Kubernetes node hosted on the machine by summarizing  node conditions.
+	// If the conditions defined in a Kubernetes node (i.e., NodeReady, NodeMemoryPressure, NodeDiskPressure, NodePIDPressure, and NodeNetworkUnavailable) are in a healthy state, it will be set to True.
+	MachineNodeHealthyCondition ConditionType = "NodeHealthy"
```
Should we? Doesn't NodeReady require CNI?

fabriziopandini replied:
Yes, this is debatable, however, I'm inclined to have this included because:
- installing CNI could be easily achieved during cluster provisioning via ClusterResourceSet, so I'm expecting Node ready should happen immediately after infrastructure provisioning
- in the conditions proposal we are stating 
  - A Ready condition SHOULD be provided at object level to represent the overall operational state of the component (and IMO for a machine, this includes having the node with a CNI)
  - The Ready condition MUST be based on the summary of more detailed conditions existing on the same object
  - An object SHOULD NEVER be in status Ready=True if one of the object's conditions are false or if one of the object dependents is in status Ready=False.

However, I'm testing this, so I can have real feedback on it

vincepri replied:
This is a contract change and should be discussed in the community, let's open an issue please?

fabriziopandini replied:
From my testing, this provides a nice UX for the users, with the only minus that we are not watching nodes in the remote cluster and so sometimes it takes a little to get synchronized.
Happy to demo it if it can help

vincepri replied:
Happy to have a demo, I'd still like for us to keep the discussion and proposal in a new issue outlining the contract changes for v1alpha4

fabriziopandini replied:
Let's tackle this in a separated PR

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if err := hasHealthyNodeConditions(node); err != nil {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "Unhealthy conditions: %s", err.Error())
```
The resulting message is kind of verbose
```
NAME                                                                   READY  SEVERITY  REASON                SINCE  MESSAGE
Cluster/test1                                                          True                                   4m20s
├─ControlPlane - KubeadmControlPlane/test1-control-plane               True                                   4m20s
│ └─Machine/test1-control-plane-dtxqk                                  True                                   4m21s
│   │           ├─NodeHealthy                                          False  Warning   NodeConditionsFailed  4m20s  Unhealthy conditions: Condition Ready is false on Node: test1-control-plane-dtxqk
```
What about 
- dropping `Unhealthy conditions:`, because this is redundant given that there is already  `NodeConditionsFailed` 
- dropping on Node: test1-control-plane-dtxqk given that the conditions is already linked to a machine/node; if not, at least we should avoid to repeat the node name for each failing condition

fabriziopandini commented:
Report from testing:
- Changes (with the additional changes in the above comments) seems good to me
- When deleting a machine, we should set NodeHealthy to false, reason Deleting

sedefsavas commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+
+func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	logger := r.Log.WithValues("machine", machine.Name, "namespace", machine.Namespace)
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return nil
+	}
```
Yes this is redundant.

sedefsavas commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+			}
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
```
Should we do the same if ProviderID is missing?

sedefsavas replied:
Actually, once ProviderID is there and we cannot get the node due to `ErrNodeNotFound`, it must have already passed provisioning. Because here `getNode()` is using providerID.

sedefsavas replied:
Opened this PR for cleaning up RequeueAfterError so that we can remove it here too: https://github.com/kubernetes-sigs/cluster-api/pull/3723

vincepri commented:
/milestone v0.3.11

sedefsavas commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,150 @@
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
```
That is what I thought initially but then we are skipping provisioning state and showing conditions on a provisioning Node if we do this. Because at the beginning of this function, it will reconcile until Machine gets a providerID, and tries to assign nodeRef here once it gets the providerID. Skipping patching conditions in this reconcile and waiting the next would be better to start showing Node conditions so that we can signal the users that the node is in provisioning state.

fabriziopandini replied:
What confuses me is that at this stage the node already exists (node!=nil), so we are definitely late for reporting NodeProvisioning

IMO we should set NodeProvisioning up in the code when node == nil && machine.Status.NodeRef == nil

sedefsavas replied:
I think for a node to be in provisioning state, it should exist. If node is missing, we are reporting `NodeNotFoundReason`.
When the noderef is still missing and node has just become available is when the node is getting ready. 

Do you think we should remove `NodeNotFoundReason` then?

fabriziopandini replied:
I think we can differentiate the two lifecycle phases.
A node is provisioning if machine.Status.NodeRef == nil and node ==nil
A node is not found if machine.Status.NodeRef != nil (the node existed) and node ==nil

sedefsavas replied:
Done.

fabriziopandini commented on controllers/machine_controller.go:
```diff
@@ -604,6 +621,12 @@ func (r *MachineReconciler) reconcileDeleteExternal(ctx context.Context, m *clus
 				"failed to delete %v %q for Machine %q in namespace %q",
 				obj.GroupVersionKind(), obj.GetName(), m.Name, m.Namespace)
 		}
+		// If delete is successful, then return empty object.
+		obj, err = external.Get(ctx, r.Client, ref, m.Namespace)
+		if err != nil && !apierrors.IsNotFound(errors.Cause(err)) {
+			return nil, errors.Wrapf(err, "failed to get %s %q for Machine %q in namespace %q",
+				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+		}
```
might be I'm missing something...

> // If delete is successful, then return empty object.

At this stage I assume delete is successful, so why are we reading again the object after delete?

sedefsavas replied:
Because it actually deletes external object here but non-nil object causes requeue: 
https://github.com/kubernetes-sigs/cluster-api/blob/af6630920560ca0e12179897b96d6ea8bd830b63/controllers/machine_controller.go#L573-L576

We already know the object is gone so this is to set the obj to nil.

fabriziopandini replied:
if the problem is to set the object to nil, should we set obj == nil without calling Get?

fabriziopandini commented on controllers/machine_controller.go:
```diff
@@ -350,7 +358,15 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 		logger.Info("Deleting node", "node", m.Status.NodeRef.Name)
+		conditions.MarkFalse(m, clusterv1.MachineNodeHealthyCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "")
+		if err := patchMachine(ctx, patchHelper, m); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to patch Machine")
```
Set clusterv1.MachineNodeHealthyCondition, clusterv1.DeletionFailedReason?

fabriziopandini commented on controllers/machine_controller.go:
```diff
@@ -268,9 +266,19 @@ func (r *MachineReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
 		errs = append(errs, err)
 	}
+	if resNode.Requeue {
+		res = resNode
```
I think we should wait for https://github.com/kubernetes-sigs/cluster-api/pull/3723 to get merged before this PR so we can have cleaner handling of return values.
Also this implementation most probably is missing util.LowestNonZeroResult here

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,137 @@
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
```
This message/event seems unrelate to getNode

sedefsavas commented:
/hold
until #3723 gets merged.

sedefsavas commented on controllers/machine_controller.go:
```diff
@@ -350,7 +358,15 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
+			return ctrl.Result{}, err
+		}
 		logger.Info("Deleting node", "node", m.Status.NodeRef.Name)
+		conditions.MarkFalse(m, clusterv1.MachineNodeHealthyCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "")
```
Only concern is early signaling a node as deleting that is operational if infrastructure fails to get deleted; but this will unlikely to happen, so changed.

fabriziopandini commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,161 @@
+
+	nodeList := corev1.NodeList{}
+	for {
+		if err := c.List(context.TODO(), &nodeList, client.Continue(nodeList.Continue)); err != nil {
```
This should be a note for a future PR, but I think that searching node for `node.Spec.ProviderID` is a common pattern in CAPI, so might be worth to create an index on it...

vincepri commented:
@sedefsavas Mind rebasing? The other PR has been merged

sedefsavas commented:
/unhold

sedefsavas commented:
Rebased to release-03. We will forward-port it to the master.

sedefsavas commented:
This PR is ready to go to release-0.3.

vincepri commented:
```
sigs.k8s.io/cluster-api/controllers
  Incompatible changes:
  - ErrNodeNotFound: removed
```

Can we restore this?

vincepri commented:
/retest

vincepri commented:
@sedefsavas Can you rebase + squash? There is something going on with apidiff

sedefsavas commented:
/test pull-cluster-api-verify-release-0-3
/test pull-cluster-api-test-release-0-3

vincepri commented:
/test pull-cluster-api-test-release-0-3

vincepri commented:
/hold 

Given the comment above. We should cover deletion and proper cleanup during e2e validation

sedefsavas commented:
@fabriziopandini 
1) Thanks for this important catch. Setting the returned `obj` to nil in reconcileDeleteExternal() was causing reconcileDelete() to continue to the end, whereas it was reconciling each time an external ref gets deleted. This was not a relevant change here anyways. So removed it.
2) Added WaitingForNodeRef reason in case Machine.spec.providerID is not set yet.reconcilePhase

fabriziopandini commented:
@sedefsavas great work!
please squash commits
/hold cancel
/lgtm

fabriziopandini commented:
/lgtm

sedefsavas commented:
This was ready to go, just solved conflicts.
cc @vincepri @fabriziopandini

fabriziopandini commented:
/lgtm

vincepri commented on controllers/machine_controller.go:
```diff
@@ -355,6 +355,16 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
+	}
+	conditions.MarkFalse(m, clusterv1.MachineNodeHealthyCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "")
+	if err := patchMachine(ctx, patchHelper, m); err != nil {
+		conditions.MarkFalse(m, clusterv1.MachineNodeHealthyCondition, clusterv1.DeletionFailedReason, clusterv1.ConditionSeverityInfo, "")
```
I would expect this to fail as well if the patch doesn't go through above

sedefsavas replied:
Machines are patched in `Reconcile` after returning from `reconcileDelete`, so setting this in case it won't fail there.

vincepri commented on controllers/machine_controller.go:
```diff
@@ -367,7 +377,6 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 	// https://github.com/kubernetes-sigs/cluster-api/issues/2565
 	if isDeleteNodeAllowed {
 		logger.Info("Deleting node", "node", m.Status.NodeRef.Name)
-
```
Revert this

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,166 @@
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, failing to get that node means the node is deleted.
```
What if it's a transient error? The node might not be deleted

sedefsavas replied:
We are setting the condition and reconciling, if it is transient it will be fixed right away. Otherwise how to detect the transient error?

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, failing to get that node means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
```
This comment is a little confusing, don't we set the node reference a few lines down?

sedefsavas replied:
Before this PR we were returning directly if the NodeRef is set.
https://github.com/kubernetes-sigs/cluster-api/blob/fe28e9df0bc4dd8bb19257d79e9fa7e45f5fd22c/controllers/machine_controller_noderef.go#L45

Removed that check to be able to check the health of the node. 

And here, if node cannot be found when the NodeRef is already assigned, that means the node is deleted. But if NodeRef is not yet set, that means there was never a node before as NodeRef is nil, and hence it is not a `NodeNotFoundReason` issue.

vincepri commented on controllers/machine_controller.go:
```diff
@@ -271,7 +271,7 @@ func (r *MachineReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
 	phases := []func(context.Context, *clusterv1.Cluster, *clusterv1.Machine) (ctrl.Result, error){
 		r.reconcileBootstrap,
 		r.reconcileInfrastructure,
-		r.reconcileNodeRef,
```
Why are we removing this call but not the code?

/hold

sedefsavas replied:
There is a name change, new name is `reconcileNode `
Not removing anything.

sedefsavas replied:
We are no longer just reconciling the nodeRef, we get the node. This is why I changed the name as well.

vincepri replied:
OK sounds good 👍

vincepri reviewed (changes requested):
Need to take a bit more time to review but it seems we're potentially changing some behaviors that we should first chat about

vincepri commented on controllers/machine_controller_node.go:
```diff
@@ -0,0 +1,166 @@
+	// Set the Machine NodeRef.
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &corev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
```
Are these properly populated? Usually they're not, and they need to be filled in manually

sedefsavas replied:
This is same with what was used before:

https://github.com/kubernetes-sigs/cluster-api/blob/fe28e9df0bc4dd8bb19257d79e9fa7e45f5fd22c/controllers/machine_controller_noderef.go#L103-L107

sedefsavas commented:
I had renamed `machine_controller_noderef` to `machine_controller_node` after addressing the previous comments.
But since there are new comments, reverted it back for better visibility.

cc @vincepri

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
 	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
 	if err != nil {
-		return ctrl.Result{}, err
+		return reconcile.Result{}, err
```
These should all be `ctrl.*`

vincepri replied:
```suggestion
		return ctrl.Result{}, err
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
 )
 
-func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (reconcile.Result, error) {
```
```suggestion
func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
 	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
-		logger.Info("Machine doesn't have a valid ProviderID yet")
-		return ctrl.Result{}, nil
+		logger.Info("Cannot reconcile node, the machine doesn't have a valid ProviderID yet")
```
```suggestion
		logger.Info("Cannot reconcile Machine's Node, no valid ProviderID yet")
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
-		return ctrl.Result{}, nil
+		logger.Info("Cannot reconcile node, the machine doesn't have a valid ProviderID yet")
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.WaitingForNodeRefReason, clusterv1.ConditionSeverityInfo, "")
+		return reconcile.Result{}, nil
```
```suggestion
		return ctrl.Result{}, nil
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
 	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
 	if err != nil {
-		return ctrl.Result{}, err
+		return reconcile.Result{}, err
```
```suggestion
		return ctrl.Result{}, err
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
+				return reconcile.Result{}, errors.Wrapf(err, "no matching Node for Machine %q in namespace %q", machine.Name, machine.Namespace)
+			}
+			conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+			return reconcile.Result{Requeue: true}, nil
```
```suggestion
			return ctrl.Result{Requeue: true}, nil
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
-		return ctrl.Result{}, err
+		logger.Error(err, "Failed to retrieve Node by ProviderID")
+		r.recorder.Event(machine, corev1.EventTypeWarning, "Failed to retrieve Node by ProviderID", err.Error())
+		return reconcile.Result{}, err
```
```suggestion
		return ctrl.Result{}, err
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
+	status, message := summarizeNodeConditions(node)
+	if status == corev1.ConditionFalse {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, message)
+		return reconcile.Result{}, nil
```
```suggestion
		return ctrl.Result{}, nil
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -19,74 +19,128 @@ package controllers
+	}
+
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+	return reconcile.Result{}, nil
```
```suggestion
	return ctrl.Result{}, nil
```

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -100,19 +154,13 @@ func (r *MachineReconciler) getNodeReference(c client.Reader, providerID *nodere
 			break
 		}
 	}
-
```
Revert this

vincepri commented:
/hold cancel

fabriziopandini commented:
@sedefsavas if you squash commits, then this one can be merged

fabriziopandini commented on api/v1alpha3/condition_consts.go:
```diff
@@ -127,3 +124,24 @@ const (
+	MachineNodeHealthyCondition ConditionType = "NodeHealthy"
+
+	// WaitingForNodeRefReason (Severity=Info) documents a machine.spec.providerId is not assigned yet.
+	WaitingForNodeRefReason = "WaitingForNodeRef"
```
nit: what about 
```suggestion
	WaitingForNodeProvisioningReason = "WaitingForNodeProvisioning"
```
NodeRef is a bit too technical for a user that does not know CAPI internals

sedefsavas replied:
There is already a `NodeProvisioningReason`, I think it will be confusing.
`WaitingForNodeRefReason ` is waiting for `machine.Spec.ProviderID` and
`NodeProvisioningReason` is waiting for `machine.Status.NodeRef` when there is no node found.

fabriziopandini commented on controllers/machine_controller_noderef.go:
```diff
@@ -64,29 +54,93 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
+		switch condition.Type {
+		case corev1.NodeMemoryPressure, corev1.NodeDiskPressure, corev1.NodePIDPressure:
+			if condition.Status != corev1.ConditionFalse {
+				message += fmt.Sprintf("Condition %s is %s", condition.Type, condition.Status) + ". "
```
nit
```suggestion
				message += fmt.Sprintf("Node condition %s is %s", condition.Type, condition.Status) + ". "
```

fabriziopandini commented:
/approve

fabriziopandini commented:
/lgtm
